vtysh: fork() on boot

qlyoung opened the pull request:
When using -b flag to apply config to all running daemons, fork a copy of vtysh for each daemon we need to configure instead of doing them one at a time. This is about N times faster when you have N daemons.

Motivation for this work is that we noticed a particularly degenerate case with an...interesting configuration file that is 30,000 lines long. Applying this config with `vtysh -b` took about 10 seconds when just one daemon was running, but with several running it took close to 60 seconds. Currently vtysh is structured so that it reads each config line and sends it to each interested daemon one at a time, waiting for their response before moving to the next one, like so:

```
for line in config:
  for daemon in daemons:
    send(daemon, line)
    response = read(daemon)
```

In the boot case, our objective is to dump the entire config to all daemons as fast as possible and then exit. For this objective there's no reason that we can't just fork off a child to handle each daemon, ensuring that the daemons are never idle while their cohorts are behind fed config like so:

```
let my_daemon;

for daemon in daemons:
  my_daemon = daemon;
  if (fork() == child)
    break

if (parent)
  wait_for_all_children()
else
  apply_config(my_daemon)
```

Doing this results in a significant performance improvement, numbers follow.

---

Config
```
# wc -l /etc/frr/frr.conf
30730 /etc/frr/frr.conf
```

Daemons
```
root@archlinux /h/vagrant# ps -fu frr
UID          PID    PPID  C STIME TTY          TIME CMD
frr        40598       1 12 17:31 ?        00:01:11 /usr/lib/frr/zebra -d
frr        40610       1 11 17:31 ?        00:01:04 /usr/lib/frr/bgpd -d
frr        40627       1  0 17:31 ?        00:00:00 /usr/lib/frr/staticd -d
frr        40779       1  0 17:40 ?        00:00:00 /usr/lib/frr/ospfd -d
frr        40801       1  0 17:40 ?        00:00:00 /usr/lib/frr/ospf6d -d
```

### No fork

`-s/--single` is the new option to force the old behavior of not forking.

```
root@archlinux /h/vagrant# time vtysh -sb
________________________________________________________
Executed in   66.75 secs    fish           external
   usr time    2.28 secs  770.00 micros    2.28 secs
   sys time    2.56 secs  602.00 micros    2.56 secs
```

Process core saturation
![2021-09-15_17-44](https://user-images.githubusercontent.com/6827003/133514800-ec57e68f-cf8e-4054-b90b-4a614642e36e.png)

### Forking

Note the new output.

```
root@archlinux /h/vagrant# time vtysh -b
[40935|zebra] sending configuration
[40938|ospfd] sending configuration
[40939|ospf6d] sending configuration
[40941|bgpd] sending configuration
[40951|staticd] sending configuration
Waiting for children to finish applying config...
[40951|staticd] done
[40941|bgpd] done
[40939|ospf6d] done
[40938|ospfd] done
[40935|zebra] done

________________________________________________________
Executed in   18.97 secs    fish           external
   usr time    7.66 secs    0.00 millis    7.66 secs
   sys time    2.21 secs    1.54 millis    2.21 secs
```

Process saturation (`staticd` had already finished because the bulk of the config did not need to be sent to it)
![2021-09-15_17-56](https://user-images.githubusercontent.com/6827003/133515291-61a69fad-c7be-4b71-892f-ed1df234bdd5.png)

donaldsharp commented:
doc update please

qlyoung commented:
Wow, that OpenBSD compilation issue really should have been caught by Clang's static analyzer.

qlyoung commented:
@donaldsharp thanks, updated the man page (vtysh options aren't in the user docs). @choppsv1 I also added some docs for the `-t` option you added a while ago, if you could check for correctness I'd appreciate it.

idryzhov reviewed (commented):
Didn't review the functionality yet, but wouldn't it be better to add a new function to vtysh.c and call it from vtysh_main.c instead of making vtysh_client structure and array public?

qlyoung commented:
What's the drawback of making the `vtysh_client` structure and array public?

idryzhov commented:
IMHO keeping internal structures private and providing APIs to manipulate them makes the code cleaner.
More than that, `vtysh_main` is more of an initialization function and it would be cleaner if it doesn't do any real functionality.

qlyoung commented:
> IMHO keeping internal structures private and providing APIs to manipulate them makes the code cleaner.

As a general principle sure, and I argue for this all the time, but I really don't see what it's buying us here. I don't see how we are worse off by allowing any code in this relatively small program to see the names of all the clients. Seems like the obvious way to do it and I don't sense a need for any sort of separation of concerns here. Rather than asking about the general design principle I was asking why you see a need for it in this specific case.

eqvinox commented on doc/manpages/vtysh.rst:
```diff
@@ -67,6 +67,15 @@ OPTIONS available for the vtysh command:
 
    Display a usage message on standard output and exit.
 
+.. option:: -t, --timestamp
```
I'd like to conserve short-option characters, mostly to have the same character with the same meaning across as much as FRR as possible — can we ditch `-s` and `-t` here and have long options only for these two?

(`-s` is `--nl-bufsize` in zebra; `-t` is `--terminal` in all daemons)

qlyoung replied:
`-s` sure, talk to @choppsv1 about `-t`

choppsv1 replied:
I'm fine with dumping the `-t` might have to look for and fix uses in the tree.

eqvinox reviewed (changes requested):
So, uh... couldn't we just put this in the init script?

```
vtysh_b () {
	[ "$1" = "watchfrr" ] && return 0
	[ -r "$C_PATH/frr.conf" ] || return 0
	if [ -n "$1" ]; then
		"$VTYSH" `echo $nsopt` -b -d "$1"
	else
		daemon_list daemons
		for dmninst in $daemons; do
			daemon_inst "$dmninst"
			# TBD: multi-instance support
			"$VTYSH" `echo $nsopt` -b -d "$daemon" &
		done
		# no arg: wait for all background processes
		wait
	fi
}
```

qlyoung commented on vtysh/vtysh.c:
```diff
@@ -63,13 +63,6 @@ char *vtysh_pager_name = NULL;
 bool vtysh_add_timestamp;
 
 /* VTY shell client structure */
```
k

eqvinox replied:
:eyes:

eqvinox replied:
The comment is still misplaced. It's only a single comment line, but still, please fix it :slightly_smiling_face:

qlyoung commented:
> So, uh... couldn't we just put this in the init script?

That doesn't let you invoke it manually does it?

eqvinox commented:
> > So, uh... couldn't we just put this in the init script?
> 
> That doesn't let you invoke it manually does it?

No — is that an actual concern?  The primary user of `vtysh -b` is the init script?

choppsv1 commented:
> > > So, uh... couldn't we just put this in the init script?
> > 
> > 
> > That doesn't let you invoke it manually does it?
> 
> No — is that an actual concern? The primary user of `vtysh -b` is the init script?

vtysh has a `-b`, users can use it, it should work well.

qlyoung commented:
> No — is that an actual concern?
Yes

> The primary user of vtysh -b is the init script?
Not for me, I use it all the time.

What exactly is the problem with it being in vtysh?

qlyoung commented:
@eqvinox Are you still blocking this PR for any reason in particular?

qlyoung commented:
ci:rerun

choppsv1 reviewed (approved):
Doc change looks good.

choppsv1 commented:
ci:micronet

mwinter-osr commented:
ci:micronet

qlyoung commented:
ci:micronet

eqvinox commented:
@qlyoung you removed the short option but it's still in the docs & `--help` output… nix that please?

qlyoung commented:
You're quite right, fixed that. Also added a commit that breaks out the fork logic to its own spot in `vtysh_config.c` and allows it to be used for both `-b` and `-f`.

qlyoung commented:
looks like a bad run to me

qlyoung commented:
ci:rerun

rzalamena commented on vtysh/vtysh.h:
```diff
@@ -121,4 +137,17 @@ extern int user_mode;
+	int log_fd;
+};
+
+extern struct vtysh_client vtysh_client[21];
```
Isn't there a way we can avoid hard coding to `21`?

qlyoung replied:
See previous discussion. If you have suggestions I'm ears though!

rzalamena reviewed (commented):
Does this still apply with `mgmtd`?

qlyoung commented:
#13004
